File: apps/macos/Kelpie/Handlers/DOMHandler.swift
import WebKit

// swiftlint:disable line_length

/// Handles DOM queries: getDOM, querySelector, querySelectorAll, getElementText, getAttributes.
struct DOMHandler {
    let context: HandlerContext

    func register(on router: Router) {
        router.register("get-dom") { body in await getDOM(body) }
        router.register("query-selector") { body in await querySelector(body) }
        router.register("query-selector-all") { body in await querySelectorAll(body) }
        router.register("get-element-text") { body in await getElementText(body) }
        router.register("get-attributes") { body in await getAttributes(body) }
    }

    @MainActor
    private func getDOM(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        let selector = body["selector"] as? String ?? "html"
        let js = """
        (function() {
            var el = document.querySelector('\(JSEscape.string(selector))');
            if (!el) return {found: false};
            return {found: true, html: el.outerHTML, nodeCount: el.querySelectorAll('*').length + 1};
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js, tabId: tabId)
            guard result["found"] as? Bool == true else {
                return errorResponse(code: "ELEMENT_NOT_FOUND", message: "Selector not found: \(selector)")
            }
            return successResponse(["html": result["html"] ?? "", "nodeCount": result["nodeCount"] ?? 0])
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func querySelector(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        guard let selector = body["selector"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "selector is required")
        }
        let js = """
        (function() {
            var el = document.querySelector('\(JSEscape.string(selector))');
            if (!el) return {found: false};
            var rect = el.getBoundingClientRect();
            return {found: true, element: {tag: el.tagName.toLowerCase(), id: el.id || undefined, text: (el.textContent || '').trim().substring(0, 200), classes: Array.from(el.classList), rect: {x: rect.x, y: rect.y, width: rect.width, height: rect.height}, visible: rect.width > 0 && rect.height > 0}};
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js, tabId: tabId)
            return successResponse(result)
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func querySelectorAll(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        guard let selector = body["selector"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "selector is required")
        }
        let js = """
        (function() {
            var els = document.querySelectorAll('\(JSEscape.string(selector))');
            return {count: els.length, elements: Array.from(els).slice(0, 100).map(function(el) {
                var rect = el.getBoundingClientRect();
                return {tag: el.tagName.toLowerCase(), id: el.id || undefined, text: (el.textContent || '').trim().substring(0, 200), rect: {x: rect.x, y: rect.y, width: rect.width, height: rect.height}};
            })};
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js, tabId: tabId)
            return successResponse(result)
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }

    @MainActor
    private func getElementText(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        guard let selector = body["selector"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "selector is required")
        }
        let js = "(document.querySelector('\(JSEscape.string(selector))')?.textContent || '')"
        do {
            let text = try await context.evaluateJSReturningString(js, tabId: tabId)
            return successResponse(["text": text.trimmingCharacters(in: .whitespacesAndNewlines)])
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "ELEMENT_NOT_FOUND", message: "Element not found: \(selector)")
        }
    }

    @MainActor
    private func getAttributes(_ body: [String: Any]) async -> [String: Any] {
        let tabId = HandlerContext.tabId(from: body)
        guard let selector = body["selector"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "selector is required")
        }
        let js = """
        (function() {
            var el = document.querySelector('\(JSEscape.string(selector))');
            if (!el) return null;
            var attrs = {};
            for (var i = 0; i < el.attributes.length; i++) attrs[el.attributes[i].name] = el.attributes[i].value;
            return {attributes: attrs};
        })()
        """
        do {
            let result = try await context.evaluateJSReturningJSON(js, tabId: tabId)
            if result.isEmpty { return errorResponse(code: "ELEMENT_NOT_FOUND", message: "Element not found: \(selector)") }
            return successResponse(result)
        } catch {
            if let tabError = tabErrorResponse(from: error) { return tabError }
            return errorResponse(code: "EVAL_ERROR", message: error.localizedDescription)
        }
    }
}

File: apps/android/app/src/main/java/com/kelpie/browser/handlers/DOMHandler.kt
package com.kelpie.browser.handlers

import com.kelpie.browser.network.Router
import com.kelpie.browser.network.errorResponse
import com.kelpie.browser.network.successResponse

class DOMHandler(
    private val ctx: HandlerContext,
) {
    fun register(router: Router) {
        router.register("get-dom") { getDOM(it) }
        router.register("query-selector") { querySelector(it) }
        router.register("query-selector-all") { querySelectorAll(it) }
        router.register("get-element-text") { getElementText(it) }
        router.register("get-attributes") { getAttributes(it) }
    }

    private suspend fun getDOM(body: Map<String, Any?>): Map<String, Any?> {
        val selector = body["selector"] as? String ?: "html"
        val safe = selector.replace("'", "\\'")
        val js = "(function(){var el=document.querySelector('$safe');return el?{html:el.outerHTML.substring(0,50000),length:el.outerHTML.length}:{html:'',length:0};})()"
        return try {
            successResponse(ctx.evaluateJSReturningJSON(js))
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun querySelector(body: Map<String, Any?>): Map<String, Any?> {
        val selector = body["selector"] as? String ?: return errorResponse("MISSING_PARAM", "selector is required")
        val safe = selector.replace("'", "\\'")
        val js =
            "(function(){var el=document.querySelector('$safe');" +
                "if(!el)return{found:false};var r=el.getBoundingClientRect();" +
                "return{found:true,element:{tag:el.tagName.toLowerCase()," +
                "text:(el.textContent||'').trim().substring(0,200)," +
                "classes:Array.from(el.classList),id:el.id||null," +
                "rect:{x:r.x,y:r.y,width:r.width,height:r.height}}};})()"
        return try {
            val result = ctx.evaluateJSReturningJSON(js)
            if (result["found"] == true) successResponse(result) else errorResponse("ELEMENT_NOT_FOUND", "No element matches: $selector")
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun querySelectorAll(body: Map<String, Any?>): Map<String, Any?> {
        val selector = body["selector"] as? String ?: return errorResponse("MISSING_PARAM", "selector is required")
        val limit = (body["limit"] as? Int) ?: 50
        val safe = selector.replace("'", "\\'")
        val js =
            "(function(){var els=document.querySelectorAll('$safe');" +
                "return{elements:Array.from(els).slice(0,$limit).map(function(el){" +
                "var r=el.getBoundingClientRect();" +
                "return{tag:el.tagName.toLowerCase()," +
                "text:(el.textContent||'').trim().substring(0,200)," +
                "classes:Array.from(el.classList),id:el.id||null," +
                "rect:{x:r.x,y:r.y,width:r.width,height:r.height}};}),total:els.length};})()"
        return try {
            successResponse(ctx.evaluateJSReturningJSON(js))
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun getElementText(body: Map<String, Any?>): Map<String, Any?> {
        val selector = body["selector"] as? String ?: return errorResponse("MISSING_PARAM", "selector is required")
        val safe = selector.replace("'", "\\'")
        val js = "(function(){var el=document.querySelector('$safe');if(!el)return null;return{text:(el.innerText||el.textContent||'').trim(),html:el.innerHTML.substring(0,5000)};})()"
        return try {
            val result = ctx.evaluateJSReturningJSON(js)
            if (result.isEmpty()) errorResponse("ELEMENT_NOT_FOUND", "No element matches: $selector") else successResponse(result)
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }

    private suspend fun getAttributes(body: Map<String, Any?>): Map<String, Any?> {
        val selector = body["selector"] as? String ?: return errorResponse("MISSING_PARAM", "selector is required")
        val safe = selector.replace("'", "\\'")
        val js =
            "(function(){var el=document.querySelector('$safe');" +
                "if(!el)return null;var attrs={};" +
                "for(var a of el.attributes){attrs[a.name]=a.value;}" +
                "return{attributes:attrs,count:el.attributes.length};})()"
        return try {
            val result = ctx.evaluateJSReturningJSON(js)
            if (result.isEmpty()) errorResponse("ELEMENT_NOT_FOUND", "No element matches: $selector") else successResponse(result)
        } catch (e: Exception) {
            errorResponse("EVAL_ERROR", e.message ?: "Unknown error")
        }
    }
}
